kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,847 @@
+/// `engine_data` _must_ be processed to complete the scan. Non-selected rows
+/// _must_ be ignored.
+pub(crate) fn table_changes_action_iter(
+    engine: &dyn Engine,
+    commit_files: impl IntoIterator<Item = ParsedLogPath>,
+    table_schema: SchemaRef,
+    predicate: Option<ExpressionRef>,
+) -> DeltaResult<impl Iterator<Item = DeltaResult<TableChangesScanData>>> {
+    let json_handler = engine.get_json_handler();
+    let filter = DataSkippingFilter::new(engine, &table_schema, predicate);
+
+    let expression_handler = engine.get_expression_handler();
+    let result = commit_files
+        .into_iter()
+        .map(move |commit_file| -> DeltaResult<_> {
+            let mut scanner = LogReplayScanner::new(
+                commit_file,
+                json_handler.clone(),
+                filter.clone(),
+                expression_handler.clone(),
+                table_schema.clone(),
+            );
+            scanner.prepare_phase()?;
+            scanner.into_scan_batches()
+        })
+        .flatten_ok()
+        .map(|x| x?);
+    Ok(result)
+}
+
+// Gets the expression for generating the engine data in [`TableChangesScanData`].
+//
+// TODO: This expression is temporary. In the future it will also select `cdc` and `remove` actions
+// fields.
+fn get_add_transform_expr() -> Expression {
+    Expression::Struct(vec![
+        column_expr!("add.path"),
+        column_expr!("add.size"),
+        column_expr!("add.modificationTime"),
+        column_expr!("add.stats"),
+        column_expr!("add.deletionVector"),
+        Expression::Struct(vec![column_expr!("add.partitionValues")]),
+    ])
+}
+
+/// Processes a single commit file from the log to generate an iterator of [`TableChangesScanData`].
+/// The scanner operates in two phases that _must_ be performed in the following order:
+/// 1. Prepare phase [`LogReplayScanner::prepare_phase`]: This performs one iteration over every
+///    action in the commit. In this phase, we do the following:
+///     - Find the timestamp from a `CommitInfo` action if it exists. These are generated when
+///       In-commit timestamps is enabled.
+///     - Ensure that reading is supported on any protocol updates
+///     - Ensure that Change Data Feed is enabled for any metadata update. See  [`TableProperties`]
+///     - Ensure that any schema update is compatible with the provided `schema`. Currently, schema
+///       compatibility is checked through schema equality. This will be expanded in the future to
+///       allow limited schema evolution.
+///     - Determine if there exists any `cdc` actions.
+///     - Constructs the remove deletion vector map from paths belonging to `remove` actions to the
+///       action's corresponding [`DvInfo`]. This map will be filtered to only contain paths that
+///       exists in another `add` action _within the same commit_. This corresponds to `remove_dv`
+///       in [`TableChangesScanData`].
```

Comment:
You'll see in the full comment that I actually split the bullets into the two phases. Everything that you highlighted here is part of phase 1. Right below it you see the description of phase 2: 
```
/// 2. Scan file generation phase [`LogReplayScanner::into_scan_batches`]: This performs another
///    iteration over every action in the commit, and generates [`TableChangesScanData`]. It does
///    so by transforming the actions using [`get_add_transform_expr`], and generating selection
///    vectors with the following rules:
///     - If a `cdc` action was found in the prepare phase, only `cdc` actions are selected
///     - Otherwise, select `add` and `remove` actions. Note that only `remove` actions that do not
///       share a path with an `add` action are selected.
```

> Are we doing those checks up front for some kind of efficiency reasons? 

It'd be ideal to fail on the first phase instead of waiting all the way till 2nd phase.

Reply:
And to make sure I'm addressing everything. We definitely need the timestamp in phase 1 because phase 2 will start transforming each engine data batch with the timestamp. Thus we need to know the timestamp ahead of phase 2.

Reply:
updated. take a look